File: Grocery_Sense/GrocerySense.Core/Services/BudgetService.cs
using GrocerySense.Data;
using GrocerySense.Data.Repositories;
using GrocerySense.Domain;

namespace GrocerySense.Core;

// Port of reference-python/.../services/budget_service.py — this month's spend vs budget + spend trend.
// Gas-cost get/save are dropped (the optimizer redesign cut distance/gas; the field is never surfaced).
// Dict return replaced with the typed BudgetStatus record.
public sealed class BudgetService
{
    private readonly ConfigStore _config;
    private readonly SqliteConnectionFactory _factory;

    public BudgetService(ConfigStore config, SqliteConnectionFactory factory)
    {
        _config = config;
        _factory = factory;
    }

    public BudgetStatus GetBudgetStatus()
    {
        var month = CurrentYearMonth();
        MonthSpend spend;
        using (var conn = _factory.Open()) spend = ReceiptsRepo.GetMonthSpend(conn, month);

        // Month-end projection: current pace held for the rest of the month.
        // ponytail: naive linear extrapolation; add weekday/pay-cycle weighting only if it demonstrably misleads.
        var now = DateTime.UtcNow;
        var daysInMonth = DateTime.DaysInMonth(now.Year, now.Month);
        var projected = decimal.Round(spend.Total / now.Day * daysInMonth, 2); // now.Day is 1..31, never 0

        // ConfigStore already normalizes MonthlyBudget to null-or-positive.
        var budget = _config.Load().MonthlyBudget is { } b ? (decimal)b : (decimal?)null;
        if (budget is null)
            return new BudgetStatus(month, spend.Total, spend.ReceiptCount, null, null, null, null, "unset",
                projected, "unset");

        var remaining = budget.Value - spend.Total;
        var pctUsed = budget.Value > 0 ? (double)(spend.Total / budget.Value) : 0.0;
        var status = pctUsed > 1.0 ? "over" : pctUsed >= 0.85 ? "warning" : "ok";
        var projStatus = Grade(projected, budget.Value);
        return new BudgetStatus(month, spend.Total, spend.ReceiptCount, budget, remaining, pctUsed,
            remaining < 0, status, projected, projStatus);
    }

    // Grade a dollar amount against the budget with the same thresholds as current-spend status.
    private static string Grade(decimal amount, decimal budget)
    {
        if (budget <= 0) return "ok";
        var pct = (double)(amount / budget);
        return pct > 1.0 ? "over" : pct >= 0.85 ? "warning" : "ok";
    }

    // Monthly spend for the last N months (oldest first).
    public IReadOnlyList<SpendTrendPoint> GetTrend(int months = 12)
    {
        using var conn = _factory.Open();
        return ReceiptsRepo.GetSpendTrend(conn, months);
    }

    // Year-over-year context line for the Budget page (Stage 4 I3). Compares this month's spend to the same
    // month last year (matched by year-month key — GetTrend is sparse, skipping receipt-free months) and
    // surfaces the current-year food-inflation rate. Both months must have receipts, else EnoughHistory=false
    // and nulls — an honest empty state, never a padded number. With ~4 receipts/month the prior-year month
    // often won't exist yet; it fills in with normal use.
    public InflationContext GetInflationContext()
    {
        var now = DateTime.UtcNow;
        var currentKey = now.ToString("yyyy-MM");
        var lastYearKey = now.AddYears(-1).ToString("yyyy-MM");

        var byMonth = GetTrend(13).ToDictionary(p => p.Month, p => p.Total);
        if (!byMonth.TryGetValue(currentKey, out var current) || !byMonth.TryGetValue(lastYearKey, out var prior)
            || prior <= 0)
            return new InflationContext(null, null, EnoughHistory: false);

        var spendYoyPct = (double)((current - prior) / prior) * 100.0;

        var rates = _config.Load().FoodInflationByYear;
        double? foodPct = rates is not null && rates.TryGetValue(now.Year.ToString(), out var r) ? r : null;

        return new InflationContext(spendYoyPct, foodPct, EnoughHistory: true);
    }

    // Persist a new monthly budget; null/non-positive clears it.
    public void SaveMonthlyBudget(double? amount)
    {
        var cfg = _config.Load();
        _config.Save(cfg with { MonthlyBudget = amount is { } a && a > 0 ? a : null });
    }

    private static string CurrentYearMonth() => DateTime.UtcNow.ToString("yyyy-MM");
}

File: reference-python/src/Grocery_Sense/services/budget_service.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from Grocery_Sense.config import config_store
from Grocery_Sense.data.repositories import receipts_repo


def _current_year_month() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m")


def get_budget_status() -> Dict[str, Any]:
    """
    Return this month's spend vs the configured budget.

    Keys:
      month          - 'YYYY-MM'
      spent          - float, sum of receipt totals this month
      receipt_count  - int
      budget         - float | None  (None means not set)
      remaining      - float | None
      pct_used       - float | None  (0–1+)
      over_budget    - bool | None
      status         - 'ok' | 'warning' | 'over' | 'unset'
    """
    month = _current_year_month()
    row = receipts_repo.get_month_spend(month)
    spent = row["total"]
    receipt_count = row["receipt_count"]

    cfg = config_store.load_config()
    budget = cfg.monthly_budget

    if budget is None:
        return {
            "month": month,
            "spent": spent,
            "receipt_count": receipt_count,
            "budget": None,
            "remaining": None,
            "pct_used": None,
            "over_budget": None,
            "status": "unset",
        }

    remaining = budget - spent
    pct_used = spent / budget if budget > 0 else 0.0
    if pct_used > 1.0:
        status = "over"
    elif pct_used >= 0.85:
        status = "warning"
    else:
        status = "ok"

    return {
        "month": month,
        "spent": spent,
        "receipt_count": receipt_count,
        "budget": budget,
        "remaining": remaining,
        "pct_used": pct_used,
        "over_budget": remaining < 0,
        "status": status,
    }


def get_trend(months: int = 12) -> List[Dict[str, Any]]:
    """Return monthly spend for the last N months (oldest first)."""
    return receipts_repo.get_spend_trend(months=months)


def save_monthly_budget(amount: Optional[float]) -> None:
    """Persist a new monthly budget (None to clear)."""
    cfg = config_store.load_config()
    cfg.monthly_budget = amount if (amount is not None and amount > 0) else None
    config_store.save_config(cfg)


def save_gas_cost_per_km(rate: float) -> None:
    """Persist gas cost per km (must be positive)."""
    if rate <= 0:
        raise ValueError(f"gas_cost_per_km must be positive, got {rate}")
    cfg = config_store.load_config()
    cfg.gas_cost_per_km = float(rate)
    config_store.save_config(cfg)


def get_gas_cost_per_km() -> float:
    return config_store.load_config().gas_cost_per_km
